cmd/carpenter/internal/parse/parser.go

```diff
@@ -63,73 +70,153 @@ func RegisterDataFilter(name string, df DataFilter) {
+		s = strings.TrimSpace(s)
+		return s
+	} else if logType == LogTypeJSON {
+		return s
+	} else {
+		panic(fmt.Sprintf("SanitizeString: unknown log type %s", logType))
+	}
 }
 
-func ParseLine(line string) (map[string]interface{}, error) {
-	var obj map[string]interface{}
-	dec := json.NewDecoder(strings.NewReader(line))
-	err := dec.Decode(&obj)
-	if err != nil {
-		return nil, fmt.Errorf("could not decode line from JSON (%s): %w", line, err)
-	}
+func ParseLine(line, logType string) (*Data, error) {
+	if logType == LogTypeJSON {
+		var obj map[string]interface{}
+		dec := json.NewDecoder(strings.NewReader(line))
+		err := dec.Decode(&obj)
+		if err != nil {
+			return nil, fmt.Errorf("could not decode line from JSON (%s): %w", line, err)
+		}
 
-	return obj, nil
-}
+		var data Data
+		if err := json.Unmarshal([]byte(line), &data); err != nil {
+			return nil, fmt.Errorf("unparsable line: %w", err)
+		}
 
-func Filter(line string) (*Data, error) {
-	line = sanitizeString(line)
+		// This isn't ours.
+		if data.IsEmpty() {
+			return nil, nil
+		}
+
+		data.RawLoggerFields = obj
+
+		return &data, nil
+	} else if logType == LogTypeMixed {
+		// log line looks like this:
+		// logger.go:146: 2025-01-17T13:51:40.330+0200 INFO    CCIPCommitPlugin.evm.1337.3379446385462418246.0x075f98f19ef9873523cde0267ab8b0253904363e        commit/plugin.go:80     creating new plugin instance    {"version": "unset@unset", "plugin": "Commit", "oracleID": 1, "donID": 2, "configDigest": "000a7d1df8632e2b3479350dcca1ee46eeec889dc37eb2ab094e63a1820ba291", "p2pID": "12D3KooWBD42agWRU3khVJwYQTXnr5uG1Qmh5n1Hm2q6RUFaJRhu"}
+		// From the JSON object we can get all the log fields
+		// From the remainer of the log we can get:
+		// - timestamp
+		// - level
+		// - logger name
+		// - caller
+		// - message
+		// write a regular expression for this
+		re := regexp.MustCompile(`(?P<timestamp>\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}(?:\+\d{4})?)\s+(?P<level>\w+)\s+(?P<logger>\S+)\s+(?P<caller>\S+)\s+(?P<message>.*?)\s+(?P<jsonFields>{.*})`)
+		matches := re.FindAllStringSubmatch(line, -1)
+		if len(matches) == 0 {
+			return nil, fmt.Errorf("could not parse line: %s", line)
+		}
 
+		// extract all the named matches
+		obj := make(map[string]string)
+		for _, match := range matches {
+			for i, name := range re.SubexpNames() {
+				if i != 0 && name != "" {
+					obj[name] = match[i]
+				}
+			}
+		}
+
+		// assert that all named matches have been found
+		if len(obj) != 6 {
+			return nil, fmt.Errorf("could not parse line: %s", line)
+		}
+
+		// construct the data object
+		var data Data
+
+		// some fields in the mixed log need to be manually populated in the Data struct
+		parsedTs, err := parseCustomLayout(obj["timestamp"])
+		if err != nil {
+			return nil, fmt.Errorf("could not parse timestamp: %w", err)
+		}
+		data.Timestamp = parsedTs
+		data.Level = obj["level"]
+		data.LoggerName = obj["logger"]
+		data.Caller = obj["caller"]
+		data.Message = obj["message"]
```

Comment:
This should all happen automatically with json unmarshalling and the codec tags

Reply:
`obj` here is not a JSON object, perhaps bad naming. Its just a regex capture name -> value map. Only one of the captures (`jsonFields`) is the JSON object that we can unmarshal into `Data` (I do it below)

Reply:
Would removing that json flag from `go test` let us avoid the special handling?

Reply:
Ah no the `json` tag from `go test` will output the entire set of logs in JSON format, this `mixed` format is w/out `-json`.